File: android/src/main/java/com/margelo/nitro/broadcastevent/BroadcastEvent.kt
package com.margelo.nitro.broadcastevent

import com.facebook.proguard.annotations.DoNotStrip

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import com.facebook.react.bridge.*


@DoNotStrip
class BroadcastEvent : HybridBroadcastEventSpec() {

    companion object {
        const val DEFAULT_CATEGORY = "android.intent.category.DEFAULT"
    }

    private val registeredReceivers = mutableListOf<BroadcastReceiver>()

    /**
     * Prepare and send broadcast event data
     */
    private fun displayScanResult(intent: Intent, actionNames: List<String>, eventName: String) {
        val eventData = WritableNativeMap()

        for (actionName in actionNames) {
            val value = intent.getStringExtra(actionName)
            if (value != null) {
                eventData.putString(actionName, value)
            } else {
                logDebug("displayScanResult", "Key '$actionName' has no data.")
            }
        }

        sendEvent(eventName, eventData)
        logDebug("displayScanResult", "Event '$eventName' sent with data: $eventData")
    }

    /**
     * Unregister a broadcast receiver
     */
    override fun unregister(index: Int, promise: Promise) {
        if (index in registeredReceivers.indices) {
            val receiver = registeredReceivers[index]
            context.unregisterReceiver(receiver)
            registeredReceivers.removeAt(index)
            promise.resolve(true)
            logDebug("unregister", "Receiver at index $index unregistered successfully.")
        } else {
            val errorMessage = "Invalid index or no receivers registered."
            logError("unregister", errorMessage)
            promise.reject("UNREGISTER_ERROR", errorMessage)
        }
    }

    /**
     * Register a broadcast receiver
     */
    override fun register(filterName: String, actionNames: String, eventName: String, category: String, promise: Promise) {
        try {
            val filter = IntentFilter(filterName).apply {
                addCategory(if (category.isEmpty()) DEFAULT_CATEGORY else category)
            }
            val actions = actionNames.split(";")
            val receiver = object : BroadcastReceiver() {
                override fun onReceive(context: Context, intent: Intent) {
                    displayScanResult(intent, actions, eventName)
                }
            }

            context.registerReceiver(receiver, filter)
            registeredReceivers.add(receiver)
            promise.resolve(registeredReceivers.indexOf(receiver))
            logDebug("register", "Receiver registered with filter: '$filterName', category: '${filter.categories}'.")
        } catch (exception: Exception) {
            val errorMessage = "Failed to register receiver: ${exception.localizedMessage}"
            logError("register", errorMessage)
            promise.reject("REGISTER_ERROR", errorMessage)
        }
    }

    /**
     * Send a broadcast
     */
    override fun sendBroadcast(actionName: String, key: String, value: String, category: String, promise: Promise) {
        try {
            val intent = Intent(actionName).apply {
                putExtra(key, value)
                addCategory(if (category.isEmpty()) DEFAULT_CATEGORY else category)
            }

            context.sendBroadcast(intent)
            promise.resolve(true)
            logDebug("sendBroadcast", "Broadcast sent: action: '$actionName', key: '$key', value: '$value', category: '${intent.categories}'.")
        } catch (e: Exception) {
            val errorMessage = "Error sending broadcast: ${e.localizedMessage}"
            logError("sendBroadcast", errorMessage)
            promise.reject("SEND_BROADCAST_ERROR", errorMessage)
        }
    }
}

File: ios/BroadcastEvent.swift
import Foundation

class BroadcastEvent: HybridBroadcastEventSpec {
    private var registeredReceivers: [String] = [] // List of registered receivers
    private var actionNameList: [String] = [] // List of action names for filtering
    private var currentEventName: String = ""
    private let defaultCategory = "default_category" // Default broadcast category on iOS

    /**
     * Override to return supported event names
     */
    override func supportedEvents() -> [String]! {
        return registeredReceivers
    }

    /**
     * Extract and send event data via broadcasts
     * @param notification Notification containing event data
     */
    @objc func displayScanResult(notification: Notification) {
        var message: [String: String] = [:]

        for actionName in actionNameList {
            if let value = notification.userInfo?[actionName] as? String {
                message[actionName] = value
            } else {
                logDebug(tag: "displayScanResult", message: "Key '\(actionName)' has no data.")
            }
        }

        sendEvent(withName: currentEventName, body: message)
        logDebug(tag: "displayScanResult", message: "Event '\(currentEventName)' sent with data: \(message).")
    }

    /**
     * Unregister a broadcast receiver
     * @param index Index of the receiver to remove
     * @param resolve Promise resolved upon success
     * @param reject Promise rejected upon failure
     */
    @objc(unregister:withResolver:withRejecter:)
    public func unregister(index: Int, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        guard index >= 0 && index < registeredReceivers.count else {
            let errorMessage = "Invalid index or no registered receivers."
            logError(tag: "unregister", message: errorMessage)
            reject("UNREGISTER_ERROR", errorMessage, nil)
            return
        }

        let receiver = registeredReceivers[index]
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name(receiver), object: nil)
        registeredReceivers.remove(at: index)

        resolve(true)
        logDebug(tag: "unregister", message: "Receiver '\(receiver)' unregistered successfully.")
    }

    /**
     * Register a broadcast receiver
     * @param filterName Name used for filtering the broadcasts
     * @param actionNames Semicolon-separated list of action names to map
     * @param eventName Name of the event to emit
     * @param category Broadcast category (defaults to "default_category")
     * @param resolve Promise containing the index of the registered receiver
     * @param reject Promise rejected upon failure
     */
    @objc(register:withActionNames:withEventName:withCategory:withResolver:withRejecter:)
    public func register(filterName: String, actionNames: String, eventName: String, category: String = "", resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            let notificationName = Notification.Name(filterName + (category.isEmpty ? defaultCategory : category))
            self.actionNameList = actionNames.components(separatedBy: ";")
            self.currentEventName = eventName

            NotificationCenter.default.addObserver(
                self,
                selector: #selector(displayScanResult(notification:)),
                name: notificationName,
                object: nil
            )

            registeredReceivers.append(filterName)
            logDebug(tag: "register", message: "Receiver registered with filter: '\(filterName)', category: '\(category.isEmpty ? defaultCategory : category)'.")
            resolve(registeredReceivers.firstIndex(of: filterName))
        } catch let error {
            let errorMessage = "Failed to register receiver: \(error.localizedDescription)"
            logError(tag: "register", message: errorMessage)
            reject("REGISTER_ERROR", errorMessage, error)
        }
    }

    /**
     * Send a broadcast
     * @param actionName Name of the broadcast intent
     * @param key Key for the data in the broadcast map
     * @param value Value associated with the key in the broadcast
     * @param category Broadcast category (defaults to "default_category")
     * @param resolve Promise resolved upon success
     * @param reject Promise rejected upon failure
     */
    @objc(sendBroadcast:withKey:withValue:withCategory:withResolver:withRejecter:)
    public func sendBroadcast(actionName: String, key: String, value: String, category: String = "", resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            let notificationName = Notification.Name(actionName + (category.isEmpty ? defaultCategory : category))
            let message = [key: value]

            NotificationCenter.default.post(
                name: notificationName,
                object: nil,
                userInfo: message
            )

            resolve(true)
            logDebug(tag: "sendBroadcast", message: "Broadcast sent: action: '\(actionName)', key: '\(key)', value: '\(value)', category: '\(category.isEmpty ? defaultCategory : category)'.")
        } catch let error {
            let errorMessage = "Error sending broadcast: \(error.localizedDescription)"
            logError(tag: "sendBroadcast", message: errorMessage)
            reject("SEND_BROADCAST_ERROR", errorMessage, error)
        }
    }
}
